hexl/fft/inv-fft-avx512.cpp

```diff
@@ -0,0 +1,367 @@
+  __m512d U_imag = *X_imag;
+
+  // X = U + Y
+  *X_real = _mm512_add_pd(U_real, *Y_real);
+  *X_imag = _mm512_add_pd(U_imag, *Y_imag);
+
+  if (scalar != nullptr) {
+    __m512d v_scalar = _mm512_set1_pd(*scalar);
+    *X_real = _mm512_mul_pd(*X_real, v_scalar);
+    *X_imag = _mm512_mul_pd(*X_imag, v_scalar);
+  }
+
+  // V = U - Y
+  __m512d V_real = _mm512_sub_pd(U_real, *Y_real);
+  __m512d V_imag = _mm512_sub_pd(U_imag, *Y_imag);
+
+  // Y = V*W. Complex multiplication:
+  // (v_r + iv_b)*(w_a + iw_b) = (v_a*w_a - v_b*w_b) + i(v_a*w_b + v_b*w_a)
+  *Y_real = _mm512_mul_pd(V_real, W_real);
+  __m512d tmp = _mm512_mul_pd(V_imag, W_imag);
+  *Y_real = _mm512_sub_pd(*Y_real, tmp);
+
+  *Y_imag = _mm512_mul_pd(V_real, W_imag);
+  tmp = _mm512_mul_pd(V_imag, W_real);
+  *Y_imag = _mm512_add_pd(*Y_imag, tmp);
+}
+
+void ComplexInvT1(double_t* result_8C_intrlvd,
+                  const double_t* operand_1C_intrlvd,
+                  const double_t* W_1C_intrlvd, uint64_t m) {
+  size_t offset = 0;
+
+  // 8 | m guaranteed by n >= 16
+  HEXL_LOOP_UNROLL_4
+
+  for (size_t i = 0; i < (m >> 1); i += 8) {
+    // Referencing operand
+    const double_t* X_op_real = operand_1C_intrlvd + offset;
+
+    // Referencing result
+    double_t* X_r_real = result_8C_intrlvd + offset;
+    double_t* X_r_imag = result_8C_intrlvd + 8 + offset;
+    __m512d* v_X_r_pt_real = reinterpret_cast<__m512d*>(X_r_real);
+    __m512d* v_X_r_pt_imag = reinterpret_cast<__m512d*>(X_r_imag);
+
+    __m512d v_X_real;
+    __m512d v_X_imag;
+    __m512d v_Y_real;
+    __m512d v_Y_imag;
+
+    ComplexLoadInvInterleavedT1(X_op_real, &v_X_real, &v_X_imag, &v_Y_real,
+                                &v_Y_imag);
+
+    // Weights
+    // x =  (14r, 10r, 6r, 2r, 12r, 8r, 4r, 0r);
+    // y =  (15r, 11r, 7r, 3r, 13r, 9r, 5r, 1r);
+    __m512d v_W_real = _mm512_set_pd(
+        W_1C_intrlvd[14], W_1C_intrlvd[10], W_1C_intrlvd[6], W_1C_intrlvd[2],
+        W_1C_intrlvd[12], W_1C_intrlvd[8], W_1C_intrlvd[4], W_1C_intrlvd[0]);
+    __m512d v_W_imag = _mm512_set_pd(
+        W_1C_intrlvd[15], W_1C_intrlvd[11], W_1C_intrlvd[7], W_1C_intrlvd[3],
+        W_1C_intrlvd[13], W_1C_intrlvd[9], W_1C_intrlvd[5], W_1C_intrlvd[1]);
+    W_1C_intrlvd += 16;
+
+    ComplexInvButterfly(&v_X_real, &v_X_imag, &v_Y_real, &v_Y_imag, v_W_real,
+                        v_W_imag);
+
+    _mm512_storeu_pd(v_X_r_pt_real, v_X_real);
+    _mm512_storeu_pd(v_X_r_pt_imag, v_X_imag);
+    v_X_r_pt_real += 2;
+    v_X_r_pt_imag += 2;
+    _mm512_storeu_pd(v_X_r_pt_real, v_Y_real);
+    _mm512_storeu_pd(v_X_r_pt_imag, v_Y_imag);
+
+    offset += 32;
+  }
+}
+
+void ComplexInvT2(double_t* operand_8C_intrlvd, const double_t* W_1C_intrlvd,
+                  uint64_t m) {
+  size_t offset = 0;
+
+  // 4 | m guaranteed by n >= 16
+  HEXL_LOOP_UNROLL_4
+  for (size_t i = 0; i < (m >> 1); i += 4) {
+    double_t* X_real = operand_8C_intrlvd + offset;
+    double_t* X_imag = operand_8C_intrlvd + 8 + offset;
```

Comment:
Similar suggestion. Or you could have x_imag = (*x_real) + 8;
```suggestion
    double_t* X_imag = operand_8C_intrlvd + offset + 8;
```